sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -283,3 +590,970 @@ def get_lpmode_support(self):
+                logger.info('Read to LPL/EPL {:#x}'.format(rpl[6]))
+
+        else:
+            raise ValueError('Reply payload check code error')
+        elapsedtime = time.time()-starttime
+        logger.info('Get module FW upgrade features time: %.2f s' %elapsedtime)
+        return startLPLsize, maxblocksize, lplonly_flag, autopaging_flag, writelength
+
+    def get_module_fw_info(self):
+        """
+        This function returns firmware Image A and B version, running version, committed version
+        and whether both firmware images are valid.
+        Operational Status: 1 = running, 0 = not running
+        Administrative Status: 1=committed, 0=uncommitted
+        Validity Status: 1 = invalid, 0 = valid
+        """
+        try:
+            self.cdb
+        except AttributeError:
+            self.get_cdb_api()
+        # get fw info (CMD 0100h)
+        starttime = time.time()
+        logger.info('\nGet module FW info')
+        rpllen, rpl_chkcode, rpl = self.cdb.get_fw_info()
+        if self.cdb.cdb_chkcode(rpl) == rpl_chkcode:
+            # Regiter 9Fh:136
+            fwStatus = rpl[0]
+            ImageARunning = (fwStatus & 0x01) # bit 0 - image A is running
+            ImageACommitted = ((fwStatus >> 1) & 0x01) # bit 1 - image A is committed
+            ImageAValid = ((fwStatus >> 2) & 0x01) # bit 2 - image A is valid
+            ImageBRunning = ((fwStatus >> 4) & 0x01) # bit 4 - image B is running
+            ImageBCommitted = ((fwStatus >> 5) & 0x01)  # bit 5 - image B is committed
+            ImageBValid = ((fwStatus >> 6) & 0x01) # bit 6 - image B is valid
+
+            if ImageAValid == 0:
+                # Registers 9Fh:138,139; 140,141
+                ImageA = '%d.%d.%d' %(rpl[2], rpl[3], ((rpl[4]<< 8) | rpl[5]))
+            else:
+                ImageA = "N/A"
+            logger.info('Image A Version: %s' %ImageA)
+            
+            if ImageBValid == 0:
+                # Registers 9Fh:174,175; 176.177
+                ImageB = '%d.%d.%d' %(rpl[38], rpl[39], ((rpl[40]<< 8) | rpl[41]))
+            else:
+                ImageB = "N/A"
+            logger.info('Image B Version: %s' %ImageB)
+
+            if ImageARunning == 1:
+                RunningImage = 'A'
+            elif ImageBRunning == 1:
+                RunningImage = 'B'
+            if ImageACommitted == 1:
+                CommittedImage = 'A'
+            elif ImageBCommitted == 1:
+                CommittedImage = 'B'
+            logger.info('Running Image: %s; Committed Image: %s' %(RunningImage, CommittedImage))
+        else:
+            raise ValueError('Reply payload check code error')
+        elapsedtime = time.time()-starttime
+        logger.info('Get module FW info time: %.2f s' %elapsedtime)
+        return ImageA, ImageARunning, ImageACommitted, ImageAValid, ImageB, ImageBRunning, ImageBCommitted, ImageBValid
+
+    def module_fw_run(self, mode = 0x01):
```

Comment:
please add return status for failure/success

Reply:
done